client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -185,10 +193,35 @@ private Checkpoint completeCheckpoint(String checkpointName) {
 
     @Override
     public void resetReaderGroup(ReaderGroupConfig config) {
+        manageSubscriptions(config);
         Map<SegmentWithRange, Long> segments = getSegmentsForStreams(controller, config);
         synchronizer.updateStateUnconditionally(new ReaderGroupStateInit(config, segments, getEndSegmentsForStreams(config)));
     }
 
+    private void manageSubscriptions(ReaderGroupConfig config) {
+        ReaderGroupConfig oldConfig = getReaderGroupConfig();
+        Set<Stream> newStreams = config.isSubscriberForRetention() ? config.getStartingStreamCuts().keySet() : Collections.emptySet();
+        Set<Stream> oldStreams = oldConfig.isSubscriberForRetention() ? oldConfig.getStartingStreamCuts().keySet() : Collections.emptySet();
+
+        Set<Stream> streamsToSub = Sets.difference(newStreams, oldStreams);
+        Set<Stream> streamsToUnsub = Sets.difference(oldStreams, newStreams);
+
+        // Unsubscribe to older streams
+        streamsToUnsub.forEach(s -> getAndHandleExceptions(controller.deleteSubscriber(scope, s.getStreamName(), groupName)
+                        .exceptionally(e -> {
+                            log.warn("Failed to delete subscriber" + e);
+                            throw Exceptions.sneakyThrow(e);
```

Comment:
The application or client **must** retry failure to `deleteSubscriber` because it is absolutely critical to delete a Subscriber after this ReaderGroup has been deleted failing which CBR can get stuck at the StreamCut of this Subscriber causing the Stream to not be deleted based on Consumption